src/LanguageServer/Protocol.TestUtilities/AbstractLspMiscellaneousFilesWorkspaceTests.cs

```diff
@@ -315,26 +305,92 @@ void M()
         Assert.Equal(7, result.Single().Range.End.Character);
     }
 
-    private static async Task AssertFileInMiscWorkspaceAsync(TestLspServer testLspServer, DocumentUri fileUri)
+    [Theory, CombinatorialData]
+    public async Task TestLspTransfersFromMiscellaneousFilesToHostWorkspaceAsync(bool mutatingLspWorkspace, bool waitForWorkspace, bool fileBasedProgramContent)
+    {
+        var markup = fileBasedProgramContent ? "Console.WriteLine();" : "class C { }";
+
+        // Create a server that includes the LSP misc files workspace so we can test transfers to and from it.
+        await using var testLspServer = await CreateTestLspServerAsync(markup, mutatingLspWorkspace, new InitializationOptions { ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer });
+
+        // Include some Unicode characters to test URL handling.
+        using var tempRoot = new TempRoot();
+        var newDocumentFilePath = Path.Combine(tempRoot.CreateDirectory().Path, "ue25b\ud86d\udeac.cs");
+
+        // If this is file based, we're going to be inspecting the actual content on disk as a part of a dotnet run-api invocation
+        if (fileBasedProgramContent)
+            File.WriteAllText(newDocumentFilePath, markup);
+        var newDocumentUri = ProtocolConversions.CreateAbsoluteDocumentUri(newDocumentFilePath);
+
+        // Open the document via LSP before the workspace sees it.
+        await testLspServer.OpenDocumentAsync(newDocumentUri, "LSP text");
+
+        // Verify it is an miscellaneous document of some kind
+        var (_, miscDocument) = await GetLspWorkspaceAndDocumentAsync(newDocumentUri, testLspServer).ConfigureAwait(false);
+        Assert.NotNull(miscDocument);
+        Assert.True(await testLspServer.GetManagerAccessor().IsMiscellaneousFilesDocumentAsync(miscDocument));
+        Assert.Equal("LSP text", (await miscDocument.GetTextAsync(CancellationToken.None)).ToString());
+
+        if (waitForWorkspace)
+        {
+            // Optionaly wait for the workspace so we can test what happens if we're seeing if it's a file-based program; otherwise we can test what
```

Comment:
```suggestion
            // Optionally wait for the workspace so we can test what happens if we're seeing if it's a file-based program; otherwise we can test what
```